### ISSUE TITLE
"~const` is not allowed here" is a sub-optimal error message in one case

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust#90052]
I think this is a small diagnostic problem. This code compiles (with two warnings):

```rust
#![feature(const_fn_trait_bound)]
#![feature(const_trait_impl)]

trait HasBar {
    fn bar(&self);
}
impl const HasBar for usize {
    fn bar(&self) {}
}

struct Foo<T>(T)
where T: HasBar;

impl<T> Foo<T> where T: HasBar {
    const fn foo(&self)
    where T: ~const HasBar {
        self.0.bar();
    }
}

fn main() {}
```

(There in the impl foo I've used both HasBar and ~const HasBar, I'm not sure if this is necessary).

If you miss the 'const' from the impl foo:

```rust
#![feature(const_fn_trait_bound)]
#![feature(const_trait_impl)]

trait HasBar {
    fn bar(&self);
}
impl const HasBar for usize {
    fn bar(&self) {}
}

struct Foo<T>(T)
where T: HasBar;

impl<T> Foo<T> where T: HasBar {
    fn foo(&self)
    where T: ~const HasBar {
        self.0.bar();
    }
}

fn main() {}
```

It gives an error that doesn't suggest that a 'const' is missing in foo:

```rust
error: `~const` is not allowed here
  --> ...\test.rs:16:14
   |
16 |     where T: ~const HasBar {
   |              ^^^^^^^^^^^^^
   |
   = note: only allowed on bounds on traits' associated types and functions, const fns, const impls and its associated functions
```

**Comment by oli-obk** [comment]
hmm... so you're saying we should suggest to make the current function const?

**Comment by leonardo-m** [comment]
This is a minor diagnostic enhancement request. If you think this isn't worth doing, then we can close this down.

Currently the diagnostic says:
"only allowed on bounds on traits' associated types and functions, const fns, const impls and its associated functions"

I'd like a focused diagnostic that says something like:
"~const HasBar" isn't allowed as a bound for a non-const function."

I think less noisy error messages are better.

**Comment by fee1-dead** [comment]
fixed in #103319.
